presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeOperators.java

```diff
@@ -0,0 +1,298 @@
+                return true;
+            }
+            if (leftNull) {
+                return false;
+            }
+            return notEqual(left, right);
+        }
+
+        @LiteralParameters("p")
+        @SqlType(StandardTypes.BOOLEAN)
+        public static boolean isDistinctFrom(
+                @BlockPosition @SqlType(value = "time(p)", nativeContainerType = long.class) Block left,
+                @BlockIndex int leftPosition,
+                @BlockPosition @SqlType(value = "time(p)", nativeContainerType = long.class) Block right,
+                @BlockIndex int rightPosition)
+        {
+            if (left.isNull(leftPosition) != right.isNull(rightPosition)) {
+                return true;
+            }
+            if (left.isNull(leftPosition)) {
+                return false;
+            }
+            return notEqual(TIME.getLong(left, leftPosition), TIME.getLong(right, rightPosition));
+        }
+    }
+
+    @ScalarOperator(INDETERMINATE)
+    @LiteralParameters("p")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean indeterminate(@SqlType("time(p)") long value, @IsNull boolean isNull)
+    {
+        return isNull;
+    }
+
+    @ScalarOperator(CAST)
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("time(targetPrecision)")
+    public static long castToTime(
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        if (sourcePrecision <= targetPrecision) {
+            return time;
+        }
+
+        return round(time, (int) (MAX_PRECISION - targetPrecision)) % PICOSECONDS_PER_DAY;
+    }
+
+    @ScalarOperator(ADD)
+    @LiteralParameters({"p", "u"})
+    @SqlType("time(u)")
+    @Constraint(variable = "u", expression = "max(3, p)") // interval is currently p = 3
+    public static long timePlusIntervalDayToSecond(@SqlType("time(p)") long time, @SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long interval)
+    {
+        return add(time, interval * PICOSECONDS_PER_MILLISECOND);
+    }
+
+    @ScalarOperator(ADD)
+    @LiteralParameters({"p", "u"})
+    @SqlType("time(u)")
+    @Constraint(variable = "u", expression = "max(3, p)") // interval is currently p = 3
+    public static long intervalDayToSecondPlusTime(@SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long interval, @SqlType("time(p)") long time)
+    {
+        return timePlusIntervalDayToSecond(time, interval);
+    }
+
+    @ScalarOperator(SUBTRACT)
+    @LiteralParameters({"p", "u"})
+    @SqlType("time(u)")
+    @Constraint(variable = "u", expression = "max(3, p)") // interval is currently p = 3
+    public static long timeMinusIntervalDayToSecond(@SqlType("time(p)") long time, @SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long interval)
+    {
+        return add(time, -interval * PICOSECONDS_PER_MILLISECOND);
```

Comment:
Seems like this can return invalid time value (negative), when interval is longer than a day (eg 3 days)

Reply:
No, it can't. `add` performs the sum modulo 24h (`% PICOSECONDS_PER_DAY`)

Reply:
right, i misread the `add`